### ISSUE TITLE
question (please answer)

### ISSUE THREAD

**Issue by ahmedabbas81** [shap/shap#915]
Dear all,

In order to get the input contributions to 2 hidden layers NN output and the NN is NOT ### convolutional NN

e = shap.DeepExplainer(model, background)

Does the above line set the reference activation for every neuron in the network?

shap_values = e.shap_values(x_test[1:5])

and this line gets the contributions in the 3D list shap_values? It has the contributions of every input feature in every input instance to each of the output neurons, is that correct?

And there are no more code lines required in the code.

Do I understand correctly?
Looking forward for an answer.

Thanks

**Comment by farh33n** [comment]
As per my understanding, if your x_test has 3 features and x_test[i] is one instance then e.shap_values(x_test[1:5]) will have a list of 4 lists where each sublist contains 3 shap values where each value represents the contribution of that feature in prediction of target value of x_test[i].

**Comment by slundberg** [comment]
@farh33n almost...lists are used when the model has multiple outputs or multiple inputs. So if you have a single input tensor and a single output value you will get a numpy array back from `e.shap_values(X)` that is the same shape as `X`. If you have multiple inputs you will get a list of these arrays, one for each input tensor. If you have multiple outputs you will also get a list of explanations, one for each output. If you have both then you would have a list of lists.

@ahmedabbas81 The short answer is yes, that is all you need to do, and yes the first line sets a distribution of reference activations as specified by the background dataset you passed.

**Comment by farh33n** [comment]
@slundberg 

> @farh33n almost...lists are used when the model has multiple outputs or multiple inputs. So if you have a single input tensor and a single output value you will get a numpy array back from `e.shap_values(X)` that is the same shape as `X`. If you have multiple inputs you will get a list of these arrays, one for each input tensor. If you have multiple outputs you will also get a list of explanations, one for each output. If you have both then you would have a list of lists.
>
In case of an RNN, if shape of each input is (50,6) that is each input is a sequence of 50 instances with 6 features. The shape of shap values returned by DeepExplainer is also (50,6) but I am interested in contribution of each feature in prediction. If I sum the shap values across rows, I can get 6 values of feature importance whose sum is also equal to difference in base and predicted value. My question is that wouldn't it be nice if shape of shap values is always same as number of features? What is the intuition behind keeping it same as shape of input?

**Comment by slundberg** [comment]
@farh33n there are two reasons for the choice of always matching the input dimension. This first is that by giving the user more information and then letting them use `sum()` to collapse when they need to we given them more flexibility. Second, it requires more semantic knowledge about the meaning of different input dimensions to know when a collapse is in order. Asking for this additional knowledge would be a big design change. I have been looking into this in connection with better explainer methods for time-series data, but that is a long term thing right now.

**Comment by ahmedabbas81** [comment]
@farh33n @slundberg 
My neural network has two hidden layers and one output layer. The hidden layers have relu activation function and the output layer is sigmoid and it is multi-class classification NN.

Is using deepshap with that neural network model correct? 

I only use 
e = shap.DeepExplainer(model, background)
shap_values = e.shap_values(x_test[1:5])

as in the example in your github page. Is that correct?
Do I need to perform any processing on the contribution values that I get (I mean shap_values)? Or I can rely on them directly as the contribution values that I want.

Also, when I use softmax instead of sigmoid, I get other contribution values that are less reasonable than those I get when using sigmoid (there are important inputs that contribute poorly when using softmax). Is there a reason for that?

And a general question, can I use sigmoid as the activation function in a multi-class classification problem where the classes are mutually exclusive, and take the maximum output as the correct class?

Thanks
